File: use-cases/government/eID-demo.js
import { NovaSdk } from 'nova-sdk-js';
import * as dotenv from 'dotenv';
import { setTimeout } from 'timers/promises';
import { randomBytes } from 'crypto'; // For simulating ID file if needed

dotenv.config();

async function main() {
  // Configuration from environment
  const rpc = process.env.RPC_URL || 'https://rpc.testnet.near.org';
  const contract = process.env.CONTRACT_ID || 'nova-contract.testnet';
  const pinataKey = process.env.IPFS_API_KEY;
  const pinataSecret = process.env.IPFS_API_SECRET;
  const shadeApiUrl = process.env.SHADE_API_URL;

  const citizenPrivateKey = process.env.CITIZEN_PRIVATE_KEY;
  const citizenAccountId = process.env.CITIZEN_ACCOUNT_ID; // e.g., citizen.testnet

  const servantPrivateKey = process.env.SERVANT_PRIVATE_KEY;
  const servantAccountId = process.env.SERVANT_ACCOUNT_ID; // e.g., servant.testnet

  if (!shadeApiUrl) throw new Error('Missing SHADE_API_URL');
  if (!pinataKey || !pinataSecret || !citizenPrivateKey || !citizenAccountId || !servantPrivateKey || !servantAccountId) {
    throw new Error('Missing required env vars');
  }

  console.log('Citizen Account:', citizenAccountId);
  console.log('Public Servant Account:', servantAccountId);

  // Initialize SDK instances
  const citizenSdk = new NovaSdk(rpc, contract, pinataKey, pinataSecret, shadeApiUrl);
  await citizenSdk.withSigner(citizenPrivateKey, citizenAccountId);

  const servantSdk = new NovaSdk(rpc, contract, pinataKey, pinataSecret, shadeApiUrl);
  await servantSdk.withSigner(servantPrivateKey, servantAccountId);

  // Unique personal group ID for this citizen
  const groupId = `myID.${citizenAccountId}`;
  console.log('\n=== Step 1: Citizen creates personal group and uploads national ID once ===');

  // Create the personal group (only citizen as initial member)
  try {
    await citizenSdk.createGroup(groupId, [citizenAccountId]);
    console.log(`Personal group created: ${groupId}`);
  } catch (err) {
    console.log(`Group already exists or error: ${err.message}. Continuing...`);
  }

  // Simulate a national ID document (PDF or image bytes)
  // In a real app, this would come from file upload / camera scan
  const idDocumentData = randomBytes(1024 * 1024); // 1MB fake PDF
  const filename = 'national-id-document.pdf';

  // Upload encrypted ID document once
  const uploadResult = await citizenSdk.compositeUpload(groupId, citizenAccountId, idDocumentData, filename);
  console.log(`Encrypted national ID uploaded to IPFS: CID = ${uploadResult.cid}`);
  console.log(`On-chain hash: ${uploadResult.hash}`);

  // Wait for IPFS propagation (important in demos)
  console.log('Waiting 15s for IPFS pin propagation...');
  await setTimeout(15000);

  console.log('\n=== Step 2: Citizen grants temporary access to public servant ===');

  // Citizen adds the servant to their personal group
  await citizenSdk.addMember(groupId, servantAccountId);
  console.log(`Access granted: ${servantAccountId} added to group ${groupId}`);

  // Small delay to ensure TEE key distribution
  await setTimeout(5000);

  console.log('\n=== Step 3: Public servant retrieves and verifies the ID document ===');

  // Servant gets a fresh nonce and signed token (handled internally by SDK in real flow)
  // compositeRetrieve will automatically fetch the group key from TEE using nonce auth
  const retrieved = await servantSdk.compositeRetrieve(groupId, uploadResult.cid);

  // Verify integrity (hash match)
  const isValid = retrieved.hash === uploadResult.hash;
  console.log(`ID document retrieved successfully. Integrity valid: ${isValid}`);
  console.log(`Decrypted data length: ${retrieved.data.length} bytes`);

  if (isValid) {
    console.log('Public servant: ID verified — service can now be provided.');
  }

  console.log('\n=== Step 4: Citizen revokes access after service completion ===');

  // Citizen removes the servant from the group
  await citizenSdk.removeMember(groupId, servantAccountId);
  console.log(`Access revoked: ${servantAccountId} removed from group`);

  // Demonstrate revocation effectiveness
  console.log('Attempting retrieval after revocation (should fail or return inaccessible key)...');
  try {
    await servantSdk.compositeRetrieve(groupId, uploadResult.cid);
    console.log('Warning: Retrieval succeeded after revocation (possible TEE caching — normal in short demos)');
  } catch (err) {
    console.log('Revocation successful: Retrieval failed as expected:', err.message);
  }

  console.log('\n=== eID Use Case Demo Complete ===');
  console.log('Citizen retains full control: one-time upload, temporary access, instant revocation.');
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});

File: use-cases/government/eID-demo.rs
use nova_sdk_rs::{NovaSdk, CompositeRetrieveResult, CompositeUploadResult};
use rand::{thread_rng, RngCore};
use std::error::Error;
use std::time::Duration;
use tokio::time::sleep;
use dotenv::dotenv;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    dotenv().ok();

    // Load environment variables
    let rpc = std::env::var("RPC_URL").unwrap_or_else(|_| "https://rpc.testnet.near.org".to_string());
    let contract = std::env::var("CONTRACT_ID").unwrap_or_else(|_| "nova-contract.testnet".to_string());
    let pinata_key = std::env::var("IPFS_API_KEY")?;
    let pinata_secret = std::env::var("IPFS_API_SECRET")?;
    let shade_api_url = std::env::var("SHADE_API_URL")?;

    let citizen_private_key = std::env::var("CITIZEN_PRIVATE_KEY")?;
    let citizen_account_id = std::env::var("CITIZEN_ACCOUNT_ID")?; // e.g., citizen.testnet

    let servant_private_key = std::env::var("SERVANT_PRIVATE_KEY")?;
    let servant_account_id = std::env::var("SERVANT_ACCOUNT_ID")?; // e.g., servant.testnet

    println!("Citizen Account: {}", citizen_account_id);
    println!("Public Servant Account: {}", servant_account_id);

    // Initialize SDK for citizen and servant
    let citizen_sdk = NovaSdk::new(&rpc, &contract, &pinata_key, &pinata_secret, &shade_api_url)
        .with_signer(&citizen_private_key, &citizen_account_id)?;

    let servant_sdk = NovaSdk::new(&rpc, &contract, &pinata_key, &pinata_secret, &shade_api_url)
        .with_signer(&servant_private_key, &servant_account_id)?;

    // Unique personal group ID for this citizen
    let group_id = format!("myID.{}", citizen_account_id);
    println!("\n=== Step 1: Citizen creates personal group and uploads national ID once ===");

    // Create personal group (only citizen as initial member)
    match citizen_sdk.create_group(&group_id, &[&citizen_account_id]).await {
        Ok(_) => println!("Personal group created: {}", group_id),
        Err(e) => println!("Group may already exist or error: {}. Continuing...", e),
    }

    // Simulate a national ID document (e.g., 1MB fake PDF)
    // In a real app: read from file via std::fs::read("national-id.pdf")?
    let mut id_document_data = vec![0u8; 1024 * 1024]; // 1MB
    thread_rng().fill_bytes(&mut id_document_data);
    let filename = "national-id-document.pdf";

    // Upload encrypted ID document once
    let upload_result: CompositeUploadResult = citizen_sdk
        .composite_upload(&group_id, &citizen_account_id, &id_document_data, filename)
        .await?;
    println!("Encrypted national ID uploaded to IPFS: CID = {}", upload_result.cid);
    println!("On-chain hash: {}", upload_result.hash);

    // Wait for IPFS pin propagation
    println!("Waiting 15s for IPFS pin to propagate...");
    sleep(Duration::from_secs(15)).await;

    println!("\n=== Step 2: Citizen grants temporary access to public servant ===");

    // Citizen adds servant to their personal group
    citizen_sdk.add_member(&group_id, &servant_account_id).await?;
    println!("Access granted: {} added to group {}", servant_account_id, group_id);

    // Allow time for TEE key distribution
    sleep(Duration::from_secs(5)).await;

    println!("\n=== Step 3: Public servant retrieves and verifies the ID document ===");

    // Servant retrieves and decrypts the document via TEE-secured key
    let retrieved: CompositeRetrieveResult = servant_sdk
        .composite_retrieve(&group_id, &upload_result.cid)
        .await?;

    // Verify integrity by comparing hashes
    let is_valid = retrieved.hash == upload_result.hash;
    println!("ID document retrieved. Integrity valid: {}", is_valid);
    println!("Decrypted data length: {} bytes", retrieved.data.len());

    if is_valid {
        println!("Public servant: National ID verified — service can now be provided.");
    }

    println!("\n=== Step 4: Citizen revokes access after service completion ===");

    // Citizen removes servant from group (triggers key rotation in TEEs)
    citizen_sdk.remove_member(&group_id, &servant_account_id).await?;
    println!("Access revoked: {} removed from group", servant_account_id);

    // Demonstrate that revocation works
    println!("Attempting retrieval after revocation (should fail due to key rotation)...");
    match servant_sdk.composite_retrieve(&group_id, &upload_result.cid).await {
        Ok(_) => println!("Warning: Retrieval succeeded (possible TEE caching in short demo)"),
        Err(e) => println!("Revocation successful: Retrieval failed as expected: {}", e),
    }

    println!("\n=== eID Use Case Demo Complete ===");
    println!("Citizen maintains full sovereignty:");
    println!("- One-time encrypted upload");
    println!("- Temporary, revocable access for public servants");
    println!("- No permanent exposure of sensitive ID data");

    Ok(())
}